### PR TITLE
Added exceptions to prevent overflow bugs.

### PR DESCRIPTION
**Pull request by aliscafo**

### CODE REVIEW

File: rostime/Makefile [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+	g++ -c -Iinclude test_exceptions/autotest.cpp -o bin/autotest.o 
+
+bin/rostime_test.o: test_exceptions/rostime_test.cpp test_exceptions/rostime_test.h
+	g++ -c -Iinclude test_exceptions/rostime_test.cpp -o bin/rostime_test.o
+
+bin/duration.o: src/duration.cpp include/ros/duration.h include/ros/impl/duration.h
+	g++ -c -Iinclude src/duration.cpp -o bin/duration.o
+
+clean:
+	rm -rf bin test_main
```
Comment (dirk-thomas) [inline rostime/Makefile:25]: Please remove the added `Makefile` and add any necessary changes to the `CMakeLists.txt` instead.

File: rostime/include/ros/exception.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+#ifndef ROSLIB_EXCEPTION_H
+#define ROSLIB_EXCEPTION_H
+
+#include <stdexcept>
+
+namespace ros
+{
+
+class Exception : public std::runtime_error
```
Comment (dirk-thomas) [inline rostime/include/ros/exception.h:36]: This class seems not to be used anywhere?

File: rostime/include/ros/impl/duration.h [hunk 3]
```diff
@@ -52,6 +52,9 @@ namespace ros {
   template<class T>
   T& DurationBase<T>::fromSec(double d)
   {
+    if ((int64_t)d < INT_MIN || (int64_t)d > INT_MAX)
```
Comment (dirk-thomas) [inline rostime/include/ros/impl/duration.h:55]: Please use C++ casts here.

Same multiple times below.

File: rostime/include/ros/impl/duration.h [hunk 4]
```diff
@@ -60,6 +63,9 @@ namespace ros {
   template<class T>
   T& DurationBase<T>::fromNSec(int64_t t)
   {
+    if ((t / 1000000000) < INT_MIN || (t / 1000000000) > INT_MAX)
```
Comment (dirk-thomas) [inline rostime/include/ros/impl/duration.h:66]: Maybe adding a local variable to store the 64 bit values would make sense to avoid repeating the computation.

Same below.

File: rostime/include/ros/macros.h [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    #define ROS_HELPER_LOCAL
+#endif
+
+// Ignore warnings about import/exports when deriving from std classes.
+#ifdef _MSC_VER
+  #pragma warning(disable: 4251)
+  #pragma warning(disable: 4275)
+#endif
+
+#endif
```
Comment (dirk-thomas) [inline rostime/include/ros/macros.h:66]: Why was this file added? It doesn't seem to be used anywhere.

File: rostime/test_exceptions/autotest.cpp [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+  }
+} 
+
+
+void Test::showFinalResults() {
+  if (failedNum)
+    printf("Failed %i of %zu tests.\n", failedNum, totalNum);
+  else
+    printf("All tests passed.\n");
+}
```
Comment (dirk-thomas) [inline rostime/test_exceptions/autotest.cpp:21]: Please remove this and related files from the patch. The tests as well as their results are already being handled by CTest and the xUnit result files.

File: rostime/test_exceptions/rostime_test.cpp [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+	test_Time_sum_assignment();
+	test_Duration_constructor();
+	test_Duration_fromSec();
+	test_Duration_fromNSec();
+	test_Duration_sum();
+	test_Duration_sum_assignment();
+	test_Duration_subtract();
+	test_Duration_subtract_assignment();
+}
+
```
Comment (dirk-thomas) [inline rostime/test_exceptions/rostime_test.cpp:301]: Please use the same code style as in other files in this repo (e.g. 2 space indentation).

Same below.

**Comment by dirk-thomas** [comment]
@aliscafo Can you please address the comment in order for this PR to be considered for merging. Thank you.

**Comment by dirk-thomas** [comment]
The patch also overlaps with #58. Please take a look if that PR covers all the case which you are trying to address in this PR.

**Comment by aliscafo** [comment]
Hi Thomas,

Thank you for your answer about the pull-request. It was my university
project, so I'm not sure if my work should be considered for merging.

Regards,
Alina

19 июл. 2017 г. 1:26 пользователь "Dirk Thomas" <notifications@github.com>
написал:

**Comment by dirk-thomas** [comment]
I created #61 to cover the changes from #58. Compared to those this PR still adds a couple of sanity checks. They still seem to be valuable to have. So I separated them in #62.

**Comment by dirk-thomas** [comment]
I will close this in favor of #62 (and #61). It would be great if you could still comment here or on #62 about the modified patch. Thanks!
